app/gateway-oss/2.4.x/plugin-development/custom-logic.md

```diff
@@ -84,155 +84,125 @@ considered closed and the `:log()` function is executed.
-of them are shared between modules, like `log`):
+{{site.ce_product_name}} processes requests in **phases**. A _plugin_ is a piece of code that gets
+activated by {{site.ce_product_name}} as each phase is executed while the request gets proxied.
 
-<div class="alert alert-warning">
-  <strong>Note:</strong> Kong uses the
-  <a href="https://github.com/rxi/classic">rxi/classic</a> module to simulate
-  classes in Lua and ease the inheritance pattern.
-</div>
+Phases are limited in what they can do. For example, the `init_worker` phase
+does not have access to the `config` parameter because that information isn't
+available when {{site.ce_product_name}} is initializing each worker.
 
-```lua
--- Extending the Base Plugin handler is optional, as there is no real
--- concept of interface in Lua, but the Base Plugin handler's methods
--- can be called from your child implementation and will print logs
--- in your `error.log` file (where all logs are printed).
-local BasePlugin = require "kong.plugins.base_plugin"
+A Plugin's `handler.lua` must return a table containing the functions it must
+execute on each phase.
 
+{{site.ce_product_name}} can process HTTP and stream traffic. Some phases are executed when
+processing only when processing HTTP traffic, others when processing stream,
+and some (like `init_worker` and `log`) are invoked by both kinds of traffic.
 
-local CustomHandler = BasePlugin:extend()
+In addition to functions, a plugin must define two fields:
 
+* `VERSION` is an informative field, not used by {{site.ce_product_name}} directly. It usually
+  matches the version defined in a Plugin's Rockspec version, when it exists.
+* `PRIORITY` is used to sort Plugins before executing each of their phases.
+  Plugins with a higher priority are executed first. See the Plugin Execution Order section below
+  for more info about this field.
 
-CustomHandler.VERSION  = "1.0.0"
-CustomHandler.PRIORITY = 10
-
+The following example `handler.lua` file defines custom functions for all
+the possible phases, in both http and stream traffic. It has no functionality
+besides writing a message to the log every time a phase is invoked. Note
+that a plugin doesn't need to provide functions for all phases.
 
--- Your plugin handler's constructor. If you are extending the
--- Base Plugin handler, it's only role is to instantiate itself
--- with a name. The name is your plugin name as it will be printed in the logs.
-function CustomHandler:new()
-  CustomHandler.super.new(self, "my-custom-plugin")
-end
+```lua
+local CustomHandler = {
+  VERSION  = "1.0.0"
+  PRIORITY = 10
+}
 
 function CustomHandler:init_worker()
-  -- Eventually, execute the parent implementation
-  -- (will log that your plugin is entering this context)
-  CustomHandler.super.init_worker(self)
-
-  -- Implement any custom logic here
+  -- Implement logic for the init_worker phase here (http/stream)
+  kong.log("init_worker")
 end
```

Comment:
this sample still uses the `:` notation.  maybe it should remain in the tables above?  I agree it's much better to call them 'functions' instead of 'methods' to avoid association with inheritance and a base class, but don't see any contradiction in calling `:access()` a function instead of a method.

Reply:
I have done something intermediate - I left the `:` notation but added an extra paragraph after the code explaining what we were doing, and showing an "uncompressed" example as well.